discovery/ecs/ecs.go

```diff
@@ -0,0 +1,154 @@
+package ecs
+
+import (
+	"net"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/docker/swarm/discovery"
+)
+
+// Discovery is exported.
+type Discovery struct {
+	ecs           ecsClient
+	ec2           ec2Client
+	cluster, port string
+	heartbeat     time.Duration
+}
+
+func init() {
+	Init()
+}
+
+// Init is exported.
+func Init() {
+	discovery.Register("ecs", &Discovery{})
+}
+
+func parsePath(path string) (cluster string, port string) {
+	cluster, port, err := net.SplitHostPort(path)
+	if err != nil {
+		return path, "2375"
+	}
+	return cluster, port
+}
+
+// Initialize is exported.
+func (s *Discovery) Initialize(path string, heartbeat time.Duration, ttl time.Duration) error {
+	config := aws.DefaultConfig
+	s.heartbeat = heartbeat
+	s.ecs = ecs.New(config)
+	s.ec2 = ec2.New(config)
+	s.cluster, s.port = parsePath(path)
+	return nil
+}
+
+func (s *Discovery) fetch() (discovery.Entries, error) {
+	var entries discovery.Entries
+
+	list, err := s.ecs.ListContainerInstances(&ecs.ListContainerInstancesInput{
```

Comment:
Thanks for the tip! I'll update this with pagination support.

On Monday, July 27, 2015, Samuel Karp <notifications@github.com> wrote:

> In discovery/ecs/ecs.go
> <https://github.com/docker/swarm/pull/1069#discussion_r35614415>:
>
> > +}
> > +
> > +// Initialize is exported.
> > +func (s *Discovery) Initialize(path string, heartbeat time.Duration, ttl time.Duration) error {
> > +	config := aws.DefaultConfig
> > +	s.heartbeat = heartbeat
> > +	s.ecs = ecs.New(config)
> > +	s.ec2 = ec2.New(config)
> > +	s.cluster, s.port = parsePath(path)
> > +	return nil
> > +}
> > +
> > +func (s *Discovery) fetch() (discovery.Entries, error) {
> > +	var entries discovery.Entries
> > +
> > +	list, err := s.ecs.ListContainerInstances(&ecs.ListContainerInstancesInput{
>
> It looks like you're not paginating through the results of
> ListContainerInstances. ECS will return a maximum of 100 results
> <http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ListContainerInstances.html#ECS-ListContainerInstances-request-maxResults>
> for this call, but you can use the ListContainerInstancePages function
> <http://docs.aws.amazon.com/sdk-for-go/api/service/ecs/ECS.html#ListContainerInstancesPages-instance_method>
> to iterate through the whole set. Note that ECS will also only allow you to
> pass up to 100 Container Instances to the DescribeContainerInstances API,
> so you'd want to call that during your iteration through the results.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/swarm/pull/1069/files#r35614415>.
>